src/Enum.php

```diff
@@ -64,157 +60,237 @@ public static function __callStatic($name, $arguments)
     {
-        if (is_string($enum)) {
-            $enum = static::from($enum);
-        }
+        return $this->index;
+    }
 
-        if (! $enum instanceof $this) {
-            return false;
-        }
+    public static function getIndices(): array
+    {
+        return array_column(static::resolve(), 'index');
+    }
 
-        if ($enum->value !== $this->value) {
-            return false;
-        }
+    public function getValue(): string
+    {
+        return $this->value;
+    }
 
-        return true;
+    public static function getValues(): array
+    {
+        return array_column(static::resolve(), 'value');
     }
 
-    /**
-     * @param string[]|\Spatie\Enum\Enum[] $enums
-     *
-     * @return bool
-     */
-    public function isOneOf(array $enums): bool
+    public function isAny(array $values): bool
     {
-        foreach ($enums as $enum) {
-            if ($this->equals($enum)) {
+        foreach ($values as $value) {
+            if ($this->isEqual($value)) {
                 return true;
             }
         }
 
         return false;
     }
 
-    public function __toString(): string
+    public function isEqual($value): bool
     {
-        return $this->value;
+        if (is_int($value) || is_string($value)) {
+            $enum = static::make($value);
+        } elseif ($value instanceof $this) {
+            $enum = $value;
+        }
+
+        if (
+            isset($enum)
+            && $enum instanceof $this
+            && $enum->getValue() === $this->getValue()
+        ) {
+            return true;
+        }
+
+        return false;
     }
 
     public function jsonSerialize()
     {
-        return $this->value;
+        return $this->getValue();
+    }
+
+    public static function make($value): Enumerable
+    {
+        if (! (is_int($value) || is_string($value))) {
+            throw new TypeError(sprintf('%s::make() expects string|int as argument but %s given', static::class, gettype($value)));
+        }
+
+        $name = null;
+        $index = null;
+
+        if (is_int($value)) {
+            if (! static::isValidIndex($value)) {
+                throw new InvalidIndexException($value, static::class);
+            }
+
+            $name = array_combine(static::getIndices(), array_keys(static::resolve()))[$value];
+            $index = $value;
+            $value = array_search($index, static::toArray());
+        } elseif (is_string($value)) {
+            if (method_exists(static::class, $value)) {
+                return forward_static_call(static::class.'::'.$value);
+            }
+
+            if (static::isValidValue($value)) {
+                $index = static::toArray()[$value];
+                $name = array_combine(static::getValues(), array_keys(static::resolve()))[$value];
+            } elseif (static::isValidName($value)) {
+                $name = $value;
+                list('value' => $value, 'index' => $index) = static::resolve()[strtoupper($name)];
```

Comment:
Are you ok with using the short `[]` syntax instead of `list`?

Reply:
I would like to keep this - it's more clear and readable.

Reply:
Ok!

Reply:
Hey, adding the caches has no impact on performance. I've added a speed test and the uncached `make()` call needs the same time like the cached one.
And adding two static variables and two methods for a single call without any performance impact isn't a good trade!?